src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -2863,5 +2864,232 @@ private static bool CheckValEscape(ImmutableArray<BoundExpression> expressions,
+
+                case BoundKind.Sequence:
+                    return HasHome(((BoundSequence)expression).Value, addressKind, methodOpt, peVerifyCompatEnabled, stackLocalsOpt);
+
+                case BoundKind.AssignmentOperator:
+                    var assignment = (BoundAssignmentOperator)expression;
+                    if (!assignment.IsRef)
+                    {
+                        return false;
+                    }
+                    var lhsRefKind = assignment.Left.GetRefKind();
+                    return lhsRefKind == RefKind.Ref ||
+                        (IsReadOnly(addressKind) && lhsRefKind == RefKind.RefReadOnly);
+
+                case BoundKind.ComplexConditionalReceiver:
+                    Debug.Assert(HasHome(
+                        ((BoundComplexConditionalReceiver)expression).ValueTypeReceiver,
+                        addressKind,
+                        methodOpt,
+                        peVerifyCompatEnabled,
+                        stackLocalsOpt));
+                    Debug.Assert(HasHome(
+                        ((BoundComplexConditionalReceiver)expression).ReferenceTypeReceiver,
+                        addressKind,
+                        methodOpt,
+                        peVerifyCompatEnabled,
+                        stackLocalsOpt));
+                    goto case BoundKind.ConditionalReceiver;
+
+                case BoundKind.ConditionalReceiver:
+                    //ConditionalReceiver is a noop from Emit point of view. - it represents something that has already been pushed. 
+                    //We should never need a temp for it. 
+                    return true;
+
+                case BoundKind.ConditionalOperator:
+                    var ternary = (BoundConditionalOperator)expression;
+                    
+                    // only ref ternary may be referenced as a variable
+                    if (!ternary.IsRef)
+                    {
+                        return false;
+                    }
+ 
+                    // branch that has no home will need a temporary
+                    // if both have no home, just say whole expression has no home 
+                    // so we could just use one temp for the whole thing
+                    return HasHome(ternary.Consequence, addressKind, methodOpt, peVerifyCompatEnabled, stackLocalsOpt)
+                        && HasHome(ternary.Alternative, addressKind, methodOpt, peVerifyCompatEnabled, stackLocalsOpt);
+
+                default:
+                    return false;
+            }
+        }
+
+        /// <summary>
+        /// Special HasHome for fields. 
+        /// Fields have readable homes when they are not constants.
+        /// Fields have writeable homes unless they are readonly and used outside of the constructor.
+        /// </summary>
+        private static bool HasHome(
+            BoundFieldAccess fieldAccess,
+            AddressKind addressKind,
+            MethodSymbol methodOpt,
```

Comment:
Where is `null` ever given for `methodOpt`?

Reply:
There were a couple places, but it turned out the cause of all of them was the AwaitExpressionSpiller. That was easy to fix up, so I've done that and removed the `Opt`. Note that this helper was previously used for Emit, so a `null` method is pretty rare, but if someone ends up using this, say, before initializers are lowered, this is actually a really common case.

Reply:
> Note that this helper was previously used for Emit, so a null method is pretty rare

Where can it be null for Emit?

---
In reply to: [215768535](https://github.com/dotnet/roslyn/pull/29591#discussion_r215768535) [](ancestors = 215768535)